torch/csrc/jit/codegen/cuda/lower_predicate_elimination.cpp

```diff
@@ -450,8 +538,24 @@ class PredicateChcker : public IterVisitor {
 
       if (non_predicated_exprs_.find(input_def) !=
           non_predicated_exprs_.end()) {
-        needs_predicate_ = true;
-        return;
+        // If producer of mma is non_predicated and initialized
+        //  with the same value. The mma should not need a
```

Comment:
Why no predicate is needed in this case?

Reply:
Ideally if the producer buffer of mma is either 0 or useful value then mma doesn't need predicate. 

This still isn't yet robust and seem to only work with the pattern that's enforced in validation pass, i.e. `shared mem -> ldmatrix -> mma`. The removal with initialization analysis still needs some more extension to properly handle prolog fusions, which was part of the TODO in the comment.

Reply:
Even when a producer is properly initialized, if the expression defining the producer is not predicated, the producer may be overwritten by meaningless values. See here for the case of ReductionOp: https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/lower_predicate_elimination.cpp#L382-L384.

In the case of ReductionOp, if the non-predicated producer is also basically the same reduction with a properly initialized input, we can be sure that the non-predicated producer expression does not produce meaningless values. That's why it has this condition: https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/lower_predicate_elimination.cpp#L375.

Reply:
I see. What were the examples of initializer value overwritten by meaningless values? I don't think the supported MMA use cases without nontrivial prolog fusion can do that but I'll need to revisit the previous motivating issues to be sure.

I guess we'll have to explicitly guard against those pathological cases somehow and still remove this predicate.

Problem is that the predicates as it's been setup now cannot be applied on any MMA. The kernel would hang. That's another tech debt we could go for some other day as I'm currently only going for predicate free use cases, which either ways seems to be almost every performance critical use case.

Reply:
If it's guaranteed by the MMA usage checks, then this should be fine, but generally speaking, if a preceding expr is not predicated, a safe assumption is that we don't know what would be written to the out-of-boundary area.

Reply:
I see. Yes the lower validation currently limits the use cases and wouldn't allow complex prolog fusions. We're not yet prioritizing prolog fusion which I think would be affected by this. 

I can add this in the TODO section of the comment to track.

Reply:
Added TODO comment to track. We could revisit in a follow up especially for handling prolog fusions. Thanks for pointing this out 👍